include/aws/s3/s3_client.h

```diff
@@ -829,6 +834,46 @@ struct aws_s3_meta_request *aws_s3_client_make_meta_request(
     struct aws_s3_client *client,
     const struct aws_s3_meta_request_options *options);
 
+/**
+ * Write the next chunk of data.
+ *
+ * You must set `aws_s3_meta_request_options.send_using_async_writes` to use this function.
+ *
+ * This function is asynchronous, and returns a future (see <aws/io/future.h>).
+ * When the future completes, if there was no error, then you may call write() again.
+ *
+ * If the future completes with an error code, then write() did not succeed
+ * and you should not call it again. If the future contains any error code,
+ * the meta request is guaranteed to finish soon, if it hasn't already, so
+ * you do not need to worry about canceling the meta request after a failed write().
+ * A common error code is AWS_ERROR_S3_REQUEST_HAS_COMPLETED, indicating
+ * the meta request completed for reasons unrelated to the write() call
+ * (e.g. received a 404 error response). AWS_ERROR_INVALID_STATE usually
+ * indicates that you're calling write() incorrectly (e.g. not waiting for
+ * previous write to complete).
+ *
+ * You can wait any length of time between calls to write().
+ * Data is buffered until there is enough to upload.
+ *
+ * @param meta_request  Meta request
+ *
+ * @param data          The data to send. This data can be any size.
```

Comment:
nit: maybe mention that data behind cursor needs to be valid for duration of function call